quinn-proto/src/packet.rs

```diff
@@ -30,16 +30,35 @@ pub struct PartialDecode {
 
 #[allow(clippy::len_without_is_empty)]
 impl PartialDecode {
-    #[allow(unreachable_pub)] // fuzzing only
+    /// Create an instance of [`PartialDecode`], expecting a fixed-length local connection id
     pub fn new(
         bytes: BytesMut,
         local_cid_len: usize,
         supported_versions: &[u32],
         grease_quic_bit: bool,
+    ) -> Result<(Self, Option<BytesMut>), PacketDecodeError> {
+        Self::new_with_cid_parser(
+            bytes,
+            &FixedLengthConnectionIdParser::new(local_cid_len),
+            supported_versions,
+            grease_quic_bit,
+        )
+    }
+
+    /// Create an instance of [`PartialDecode`], with given [`ConnectionIdParser`].
+    pub fn new_with_cid_parser<Parser: ConnectionIdParser>(
```

Comment:
Same here: there are just two callers, both of which are passing in `self.local_cid_state.cid_len()`. I think we can just have one variant, and maybe implement `ConnectionIdParser` for `CidState` and `ConnectionIdGenerator`?

Reply:
I would like to keep them as is, if we really want to remove `PartialDecoder::new`, construct a parser when invoke `PartialDecoder::new_with_cid_parser`, impl ConnectionIdParser for them is less readable in my opinion.

Reply:
I'm not convinced. @Ralith any thoughts? Feels like `CidState` and `ConnectionIdGenerator` are plausible implementors of `ConnectionIdParser`.

Reply:
> I'm not convinced.

I think @thynson's comment above is out of date, and/or they decided to concede the point. The current form of the PR no longer introduces new constructors, instead tweaking the pre-existing ones.

> Feels like CidState and ConnectionIdGenerator are plausible implementors of ConnectionIdParser.

I think it would make sense for `ConnectionIdGenerator` to require `ConnectionIdParser`, yeah. We should probably take care to not perform redundant validation work for CIDs that have already been implicitly validated by finding them in the endpoint's live connection index.

This could be pursued in followup work, although we should do so before publishing this change, as it could render `FixedLengthConnectionIdGenerator` redundant.

I'm not sure `CidState` should implement `ConnectionIdParser`. It looks like it only knows about `cid_len` for the sake of judging whether CIDs are in use at all, and thereby detecting certain transport errors. User-defined CID generators might have more complex parsing requirements and, unlike `ConnectionIdGenerator`, `CidState` doesn't seem to be in a position to know about them.